File: kangtaku/second_minimum_node_in_a_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        int left = findMin(root->left, root->val);
        int right = findMin(root->right, root->val);
        if ((left == INT_MAX && right == INT_MAX) || min(left, right) == root->val) {
            return -1;
        } else {
            return min(left, right);
        }
    }
    
    int findMin(TreeNode* root, int now) {
        if (root == NULL) return INT_MAX;
        if (now < root->val) {
            return root->val;
        }
        return min(findMin(root->left, now), findMin(root->right, now));
    }
};

File: byssup/tree/Second Minimum Node In a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findSecondMinimumValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        nums = set()
        self.rec(root, nums)
        nums = list(nums)
        nums.sort()
        if len(nums) < 2:
            return -1
        return nums[1]

    def rec(self, node, nums):
        if node:
            nums.add(node.val)
            self.rec(node.left, nums)
            self.rec(node.right, nums)